src/Horus/CairoSemantics/Runner.hs

```diff
@@ -134,10 +134,12 @@ interpret = iterM exec
     eConstraints . csAsserts
       .= ( ( ExistentialAss
               ( \mv ->
+                  -- I am sorry about this code.
                   let restAss' = map (builderToAss mv . fst) restAss
                       asAtoms = concatMap (\x -> fromMaybe [x] (unAnd x)) restAss'
                       restExp' = map fst restExp
-                   in (a mv .|| Expr.not (Expr.and (filter (/= a mv) asAtoms)))
+                      isNotPre = (`notElem` fromMaybe [a mv] (unAnd (a mv)))
```

Comment:
Since we use the function `\x -> fromMaybe [x] (unAnd x)` twice, maybe we can move it into a `let` binding or `where`? :)